### PR TITLE
feat: Added an HOC to sort the models in the table

### PR DESCRIPTION
**Pull request by rav3n11**
- added a Higher order component that is designed to wrap the existing ModelsTable and provide sorting functionality
- currently the table doesn't quite take advantage of the sorting, but the defaults allow the model Import behaviour to be stable

### CODE REVIEW

File: frontend/src/modeldirectory/components/modelTables/withSorting.tsx [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+          // Determine the direction
+          return sortConfig.direction === 'ascending' ? comparison : -comparison;
+        });
+      }
+      return sortableItems;
+    }, [props.models, sortConfig]);
+
+    const requestSort = useCallback((key: SortableKeys, direction:SortDirection = key === 'name' ? SortDirection.ASCENDING : SortDirection.DESCENDING ) => {
+      if (sortConfig.key === key && sortConfig.direction === SortDirection.ASCENDING) {
```
Comment (ApostolosBenisis) [inline frontend/src/modeldirectory/components/modelTables/withSorting.tsx:37]: here add a comment about the intention, I could not figure it out by just looking at it for a couple of seconds

File: frontend/src/modeldirectory/components/modelTables/withSorting.tsx [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+      if (sortConfig !== null) {
+        sortableItems.sort((a, b) => {
+          let comparison;
+
+          if (sortConfig.key === 'updatedAt') {
+            // Compare dates by converting them to timestamps
+            comparison = a.updatedAt.getTime() - b.updatedAt.getTime();
+          } else {
+            // For strings, use localeCompare for a correct comparison
+            comparison = a[sortConfig.key].localeCompare(b[sortConfig.key]);
```
Comment (ApostolosBenisis) [inline frontend/src/modeldirectory/components/modelTables/withSorting.tsx:26]: generally speaking a[sortConfig.key] could be undefined at runtime, is there a trycatch needed?

Comment (ApostolosBenisis) [inline frontend/src/modeldirectory/components/modelTables/withSorting.tsx:26]: versions should be sorted [Semantic](https://semver.org/)

File: frontend/src/modeldirectory/components/modelTables/withSorting.tsx [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+
+const withSorting = (WrappedComponent: React.FC<ModelsTableProps>) => {
+  return (props: ModelsTableProps) => {
+    const [sortConfig, setSortConfig] = useState<SortConfig>({ key: 'updatedAt', direction: SortDirection.DESCENDING });
+
+    const sortedModels = useMemo(() => {
+      let sortableItems = [...props.models];
+      if (sortConfig !== null) {
+        sortableItems.sort((a, b) => {
```
Comment (ApostolosBenisis) [inline frontend/src/modeldirectory/components/modelTables/withSorting.tsx:18]: perhaps for better testing, especially when the sortable keys grows and for boundary conditions ( nulls, undefined) put the sorting function in a separate file  
 Then most of your tests will also move to a separate file

File: frontend/src/modeldirectory/components/modelTables/withSorting.types.ts [hunk 4]
```diff
@@ -0,0 +1,6 @@
+export type SortableKeys = 'updatedAt' | 'name' | 'version'; // Define the keys by which sorting can be done
```
[the remark this answers is not in the document]
Comment (rav3n11) [reply frontend/src/modeldirectory/components/modelTables/withSorting.types.ts:1]: Yeah but do we want that? Wouldn't we want the more recently updated models up higher?

[the remark this answers is not in the document]
Comment (rav3n11) [reply frontend/src/modeldirectory/components/modelTables/withSorting.types.ts:1]: Cool. Cool. Updating...

File: frontend/src/modeldirectory/components/modelTables/withSorting.tsx [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+          if (sortConfig.key === 'updatedAt') {
+            // Compare dates by converting them to timestamps
+            comparison = a.updatedAt.getTime() - b.updatedAt.getTime();
+          } else {
+            // For strings, use localeCompare for a correct comparison
+            comparison = a[sortConfig.key].localeCompare(b[sortConfig.key]);
+          }
+
+          // Determine the direction
+          return sortConfig.direction === 'ascending' ? comparison : -comparison;
```
[the remark this answers is not in the document]
Comment (rav3n11) [reply frontend/src/modeldirectory/components/modelTables/withSorting.tsx:30]: Oversight. Will fix.

File: frontend/src/modeldirectory/components/modelTables/withSorting.tsx [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+  return (props: ModelsTableProps) => {
+    const [sortConfig, setSortConfig] = useState<SortConfig>({ key: 'updatedAt', direction: SortDirection.DESCENDING });
+
+    const sortedModels = useMemo(() => {
+      let sortableItems = [...props.models];
+      if (sortConfig !== null) {
+        sortableItems.sort((a, b) => {
+          let comparison;
+
+          if (sortConfig.key === 'updatedAt') {
```
[the remark this answers is not in the document]
Comment (rav3n11) [reply frontend/src/modeldirectory/components/modelTables/withSorting.tsx:21]: Hey... yeah.

File: frontend/src/modeldirectory/components/modelTables/withSorting.tsx (hunk 1 above)
Comment (rav3n11) [reply frontend/src/modeldirectory/components/modelTables/withSorting.tsx:37]: I see. Will document.

File: frontend/src/modeldirectory/components/modelTables/withSorting.tsx (hunk 2 above)
Comment (rav3n11) [reply frontend/src/modeldirectory/components/modelTables/withSorting.tsx:26]: Ooh. Cool. Will add that.

File: frontend/src/modeldirectory/components/modelTables/withSorting.tsx (hunk 3 above)
Comment (rav3n11) [reply frontend/src/modeldirectory/components/modelTables/withSorting.tsx:18]: Interesting. Cool. Like the parseFiles... I can do that...

File: frontend/src/modeldirectory/components/modelTables/withSorting.tsx (hunk 2 above)
Comment (rav3n11) [reply frontend/src/modeldirectory/components/modelTables/withSorting.tsx:26]: Ooh. Cool. Will add that.

Comment (rav3n11) [reply frontend/src/modeldirectory/components/modelTables/withSorting.tsx:26]: Yup. too open. will fix.

File: frontend/src/modeldirectory/components/modelTables/withSorting.types.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (ApostolosBenisis) [reply frontend/src/modeldirectory/components/modelTables/withSorting.types.ts:1]: I am not sure, with updatedAt polling could result in "surprising behaviour". 
but i am not insisting on this one

[the remark this answers is not in the document]
Comment (rav3n11) [reply frontend/src/modeldirectory/components/modelTables/withSorting.types.ts:1]: Oh I see. We can see if the behavior is distracting and change it accordingly?

**Comment by rav3n11** [comment]
This feature was "Ahead of its time :laughing: ". It has been abandoned in favor of a simpler implementation...
